src/libraries/System.Text.Json/src/System/ReflectionExtensions.cs

```diff
@@ -42,6 +43,39 @@ public static bool IsAssignableFromInternal(this Type type, Type from)
         private static bool HasJsonConstructorAttribute(ConstructorInfo constructorInfo)
             => constructorInfo.GetCustomAttribute<JsonConstructorAttribute>() != null;
 
+        public static bool HasRequiredMemberAttribute(this ICustomAttributeProvider memberInfo)
+        {
+#if NET7_0_OR_GREATER
+            return memberInfo.IsDefined(typeof(RequiredMemberAttribute), inherit: true);
+#else
+            return memberInfo.HasCustomAttributeWithName("System.Runtime.CompilerServices.RequiredMemberAttribute");
+#endif
+        }
+
+        public static bool HasSetsRequiredMembersAttribute(this ICustomAttributeProvider memberInfo)
+        {
+#if NET7_0_OR_GREATER
+            return memberInfo.IsDefined(typeof(SetsRequiredMembersAttribute), inherit: true);
+#else
+            return memberInfo.HasCustomAttributeWithName("System.Diagnostics.CodeAnalysis.SetsRequiredMembersAttribute");
+#endif
+        }
+
+#if !NET7_0_OR_GREATER
+        private static bool HasCustomAttributeWithName(this ICustomAttributeProvider memberInfo, string name)
+        {
+            foreach (object attribute in memberInfo.GetCustomAttributes(inherit: true))
```

Comment:
Per https://github.com/dotnet/runtime/pull/72937/commits/b422a9af9dad3da55b1d0bcbaaa80f12aa4549ab#r932216827, `inherit` should probably not be hardcoded here and instead be specified explicitly in the individual attribute resolver methods.